### PR TITLE
Properly catch errors in ldflag-gathering and fail the build

### PR DESCRIPTION
**Pull request by Groxx**
The dubious-ownership issue in CI appears to be happening here too,
and might be why some builds have "unknown" as their commit SHAs.

Building this change locally properly fails to build binaries when
it cannot be gathered, and includes the data when it can:
```
VERSION:
   CLI feature version: 1.7.0
   Release version: v1.2.7-prerelease2-dirty
   Build commit: 2023-12-22T20:30:21+00:00-d4d205fe3
```

---

Previously there were basically three issues at play:
- git's new `safe.directory` thing, now applied everywhere rather than just one target
- the in-variable-expansion shell error would be swallowed and
(AFAICT) be completely undetectable elsewhere in Make.
- [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155) and similar "you captured output but you also suppressed the error code" issues in scripts.

Since fixing the "always include the version, fail build if it does not work" issue
involved discovering and fixing all three, they're all in this PR.

We should seriously consider adding shellcheck to our CI pipeline.

### CODE REVIEW

File: scripts/build-with-ldflags.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env sh
```
Comment (davidporter-id-au) [inline scripts/build-with-ldflags.sh:1]: `#!/usr/bin/env sh` I assume? og shell is hard to use/read

File: scripts/get-ldflags.sh [hunk 2]
```diff
@@ -1,5 +1,7 @@
 #!/bin/sh
```
Comment (davidporter-id-au) [inline scripts/get-ldflags.sh:1]: bash?

File: scripts/build-with-ldflags.sh (hunk 1 above)
Comment (Groxx) [reply scripts/build-with-ldflags.sh:1]: nothing in here uses bash-specific stuff, doesn't really seem necessary?

Comment (taylanisikdemir) [reply scripts/build-with-ldflags.sh:1]: I'd lean towards having the same interpreter for all .sh files just for consistency. preferably bash

**Review by taylanisikdemir — APPROVED** [review]
LGTM with comment

File: scripts/build-with-ldflags.sh (hunk 1 above)
Comment (Groxx) [reply scripts/build-with-ldflags.sh:1]: ah, yea, we're aaaaalmost consistent on `#!/bin/bash` by the look of things.  might as well continue that trend I guess.

**Comment by Groxx** [comment]
hmmm
```
++ ./scripts/get-ldflags.sh LDFLAG
fatal: detected dubious ownership in repository at '/cadence'
To add an exception for this directory, call:
	git config --global --add safe.directory /cadence
fatal: detected dubious ownership in repository at '/cadence'
To add an exception for this directory, call:
	git config --global --add safe.directory /cadence
+ FLAGS='-X github.com/uber/cadence/common/metrics.Revision= -X github.com/uber/cadence/common/metrics.Branch= -X github.com/uber/cadence/common/metrics.ReleaseVersion=unknown -X github.com/uber/cadence/common/metrics.BuildDate=2024-01-06-00:04:12 -X github.com/uber/cadence/common/metrics.BuildTimeUnix=1704499452'
+ exec go build -ldflags '-X github.com/uber/cadence/common/metrics.Revision= -X github.com/uber/cadence/common/metrics.Branch= -X github.com/uber/cadence/common/metrics.ReleaseVersion=unknown -X github.com/uber/cadence/common/metrics.BuildDate=2024-01-06-00:04:12 -X github.com/uber/cadence/common/metrics.BuildTimeUnix=1704499452' -o cadence-cassandra-tool cmd/tools/cassandra/main.go
```
apparently there are still gaps in that.  gonna hunt for a higher-level place to do it.

**Comment by Groxx** [comment]
much better:
```
+ FLAGS='-X github.com/uber/cadence/common/metrics.Revision=2024-01-05T18:34:15-06:00-07165ee57 -X github.com/uber/cadence/common/metrics.Branch=HEAD -X github.com/uber/cadence/common/metrics.ReleaseVersion=v1.2.7-prerelease2 -X github.com/uber/cadence/common/metrics.BuildDate=2024-01-06-00:40:23 -X github.com/uber/cadence/common/metrics.BuildTimeUnix=1704501623'
```
gonna clean up a bit

**Comment by Groxx** [comment]
dangit. 

> Declare and assign separately to avoid masking return values.
> See [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155).

so we should probably just add shellcheck.

**Comment by Groxx** [comment]
dunno what this implies: 
```
Unable to find image 'koalaman/shellcheck:stable' locally
stable: Pulling from koalaman/shellcheck
ee9c4b55cadb: Pull complete
7d79c83e79c5: Pull complete
Digest: sha256:f35e8987b02760d4e76fc99a68ad5c42cc10bb32f3dd2143a3cf92f1e5446a45
Status: Downloaded newer image for koalaman/shellcheck:stable
Unknown shell: warning
🚨 Error: The command exited with status 1
user command error: The plugin shellcheck command hook exited with status 1
```
and going to ignore it for now.  `make lint` finds this stuff at least, we can address it later.

File: scripts/buildkite/golint.sh [hunk 3]
```diff
@@ -4,15 +4,6 @@ set -ex
 
 make go-generate && make fmt && make lint && make copyright
 
-# buildkite-git complains about "dubious ownership",
-# i.e. user-who-created is different than user-who-statused.
-#
-# that's fine, just override it if it looks like we're in buildkite.
-# elsewhere it's probably best to not touch this, and the path is likely wrong.
-if [ -n "$BUILDKITE" ]; then
-  git config --global --add safe.directory /cadence
```
Comment (Groxx) [inline scripts/buildkite/golint.sh:13]: this is now done in the buildkite dockerfile, which appears to catch everything much more simply.

File: scripts/get-ldflags.sh [hunk 4]
```diff
@@ -8,19 +13,21 @@ if [ "$MODE" != "LDFLAG" ] && [ "$MODE" != "ECHO" ]; then
...
-  export RELEASE_VERSION=$(git describe --tags --abbrev=0 --dirty 2>/dev/null || echo unknown)
+  RELEASE_VERSION=$(git describe --tags --abbrev=0 --dirty 2>/dev/null || echo unknown)
 else
   # If passing a CADENCE_RELEASE_VERSION explicitly, then use it
-  export RELEASE_VERSION=${CADENCE_RELEASE_VERSION}
+  RELEASE_VERSION=${CADENCE_RELEASE_VERSION}
 fi
 
+# var-setting and exporting must be done separately or `set -e` does not catch errors
+export GIT_REVISION GIT_BRANCH BUILD_DATE BUILD_TS_UNIX BASE_PACKAGE RELEASE_VERSION
```
Comment (Groxx) [inline scripts/get-ldflags.sh:30]: TIL.  shellcheck catches this, I tried to add a stable version to our pipeline, but failed several times and don't feel like spending more hours figuring out the internals of buildkite :|

it'd be pretty easy to add to our makefile, but getting a stable version across local and ci is probably hard.  also probably worthwhile either way, but still.

Comment (taylanisikdemir) [reply scripts/get-ldflags.sh:30]: why do we need these exported?

**Review by taylanisikdemir — APPROVED** [review]

File: scripts/get-ldflags.sh (hunk 4 above)
Comment (Groxx) [reply scripts/get-ldflags.sh:30]: since it's just echoing: good point!  we do not.
I can just drop that in here.

File: docker/start.sh [hunk 5]
```diff
@@ -130,11 +130,11 @@ wait_for_postgres() {
 wait_for_es() {
     server=`echo $ES_SEEDS | awk -F ',' '{print $1}'`
     URL="http://$server:$ES_PORT"
-    curl -s $URL 2>&1 > /dev/null
+    curl -s $URL > /dev/null 2>&1
```
Comment (Groxx) [inline docker/start.sh:133]: TIL.  I've gotta adjust my muscle memory apparently.

**Pull request merged**